provider_client.go

```diff
@@ -215,13 +200,47 @@ func (client *ProviderClient) Request(method, url string, options RequestOpts) (
 	if !ok {
 		body, _ := ioutil.ReadAll(resp.Body)
 		resp.Body.Close()
-		return resp, &UnexpectedResponseCodeError{
+		respErr := &UnexpectedResponseCodeError{
 			URL:      url,
 			Method:   method,
 			Expected: options.OkCodes,
 			Actual:   resp.StatusCode,
 			Body:     body,
 		}
+
+		errType := options.ErrorType
+		switch resp.StatusCode {
+		case http.StatusUnauthorized:
+			if client.ReauthFunc != nil {
+				err = client.ReauthFunc()
+				if err != nil {
+					return nil, fmt.Errorf("Error trying to re-authenticate: %s", err)
+				}
+				if options.RawBody != nil {
+					options.RawBody.Seek(0, 0)
+				}
+				resp, err = client.Request(method, url, options)
+				if err != nil {
+					return nil, fmt.Errorf("Successfully re-authenticated, but got error executing request: %s", err)
+				}
+			}
+			err = defaultError401{}
```

Comment:
Oh, I'd just noticed that the `defaultError401` type definitions all embed an `*UnexpectedResponseCodeError`:

```go
type defaultError405 struct {
	*UnexpectedResponseCodeError
}
```